More plugin events

Elfangor93 opened the pull request:
This is a redone of #58.

During my work with the JoomGallery I am missing Plugin events especially during the administration tasks (upload, forms, batch processes, ...).

I am convinced that meaningful and well located plugin events are a great benefit for the component.

Erftralle commented:
Since this pull request does not only contain the code of the old pull request, it must be completely re-tested.

I'll start with the new events in JoomAdminConfig or JoomConfig:
What could the events in JoomAdminConfig or JoomConfig be used for? Please provide an explanation based on an example. Is there already a plugin that you have used that uses the events in JoomAdminConfig or JoomConfig?

Erftralle commented on administrator/components/com_joomgallery/helpers/adminconfig.php:
```diff
@@ -96,6 +96,9 @@ public function save($newconfig = null, $id = false, $existing_row = 0, $group_i
     }
 
+    JPluginHelper::importPlugin('content');
+    $this->_mainframe->triggerEvent('onContentBeforeSave', array(_JOOM_OPTION.'.config', &$config, $isNew, $data));
```
$this->_mainframe is not declared in this content, this call will throw an error

Erftralle commented on administrator/components/com_joomgallery/helpers/adminconfig.php:
```diff
@@ -113,6 +116,8 @@ public function save($newconfig = null, $id = false, $existing_row = 0, $group_i
       return false;
     }
 
+    $this->_mainframe->triggerEvent('onContentAfterSave', array(_JOOM_OPTION.'.config', &$config, $isNew));
```
$this->_mainframe is not declared in this content, this call will throw an error

Elfangor93 replied:
solved

Erftralle commented on administrator/components/com_joomgallery/helpers/adminconfig.php:
```diff
@@ -1046,6 +1051,9 @@ public function delete($id, $force = false)
     }
 
+    JPluginHelper::importPlugin('content');
+    $this->_mainframe->triggerEvent('onContentAfterDelete', array(_JOOM_OPTION.'.config', $config));
```
$this->_mainframe is not declared in this content, this call will throw an error

Erftralle commented on administrator/components/com_joomgallery/helpers/config.php:
```diff
@@ -383,6 +383,10 @@ public function __construct($id = null)
       $config->load($this->_id);
 
+      JPluginHelper::importPlugin('content');
+      $_mainframe = JFactory::getApplication('administrator');
```
JoomConfig::__construct is also called by the frontend, should the event also be triggered when loading the configuration in the frontend and for what is it usefull?

Elfangor93 replied:
JoomConfig::__construct loads the configurations from the db and offers them to the application during run with $this->_config. With a plugin using the 'onContentPrepareData' trigger in the JoomConfig::__construct method you are able to manipulate the configuration object. For example you could add another configuartion rule wich is not part of the #__joomgallery_config table without the need of changing the JoomGallery-Core.

Erftralle replied:
Please don't use an underscore as the first letter of a local variable. Just call it **mainframe** (please remember also the other occurrences where you changed ```$this->_mainframe``` to ```$_mainframe```).

Correct me if I am wrong, but the constructor is also called in the frontend application, so you should just call  
```
$mainframe = JFactory::getApplication();
```
to get the right application object, otherwise you will create a new administrator application object when calling the constructor in the frontend.

Why should I add another configuartion option in a plugin without changing the JoomGallery core which I can not use inside JoomGallery without changing the JoomGallery core?

Erftralle commented on administrator/components/com_joomgallery/helpers/config.php:
```diff
@@ -400,6 +404,9 @@ public function __construct($id = null)
       }
 
+      JPluginHelper::importPlugin('content');
+      $this->_mainframe->triggerEvent('onContentPrepareData', array(_JOOM_OPTION.'.config', (object)$properties));
```
$this->_mainframe is not declared in this content, this call will throw an error

Elfangor93 commented:
`What could the events in JoomAdminConfig or JoomConfig be used for?`
JoomGallery includes multiple content types: images, categories, comments, configuartions, nameshields, users, votes. Some of them include administrator areas where you can add, edit, delete or change state of the contents.
This is alo true for the content type configuration. There should also be the possibility to hook in with a pluign to the administration process of the configuration just as it is possible for the images and the categories.

`Is there already a plugin that you have used that uses the events in JoomAdminConfig or JoomConfig?`
No. How could there possibly be that a plugin is using an event that doesn't exist?
If there is no plugin event there will never be a plugin using it.
For plugin events I guess the rule is: If it does no good, it does no harm.

Erftralle commented:
> There should also be the possibility to hook in with a pluign to the administration process of the configuration just as it is possible for the images and the categories.

Sorry, but this does not answer my question for what these events in JoomAdminConfig or JoomConfig could be used for.

>No. How could there possibly be that a plugin is using an event that doesn't exist?
If there is no plugin event there will never be a plugin using it.

I just thought, that it could have been that you needed these events in a personal modified JoomGallery you used on a website you created. But I see that you also have not a real idea for what these events could be used for at all.

> For plugin events I guess the rule is: If it does no good, it does no harm.

Sorry, but I do not understand this rule.


Importing the plugins when loading the configuration at least costs process time and when introducing new events, I think, you should at least have a real idea for what they could be used for. This is just my personal opinion.
But I would like to let you convince me that they are really useful. 
Maybe there are more opinions.

Elfangor93 commented:
I think the point to discuss is wether we wanna offer plugin events that are not used at the moment but could be useful in future plugin development or if we wanna keep our code minimal and dont offer plugin events in advance.

In my opinion it does no harm to trigger plugin events that are not used at the moment. On the contrary, the possibility to extend the component with your own plugins helps to make the component attractive for developer and advanced joomla users. They dont want to modify the JoomGallery core first to bring in the plugin events they need. This plugin events should already be available.
In my opinion we should offer as much plugin events as possible. But of course not stupid ones ;-). This makes the component agile for customisation which I think is important for a component with such a wide range of applications.

MrMusic commented:
In my opinion, we should introduce new plugin events if it brings concrete benefits.

>During my work with the JoomGallery I am missing plugin events especially during...

It seems you already have an idea for the usage. So I'm in favor of adding the event!

On the other hand, I'm in favor of the principle "as much as necessary - as little as possible". 
Why? With every code change, isn't there a risk of catching problems somewhere?
Another aspect is the consumption of resources. Aren't these plugin events slowing down all?

>But of course not stupid ones ;-).

Difficult to say where the benefit ends and it gets stupid.
In theory, you can insert an additional plugin event for EVERY action and function. But does that make sense?

The JoomGallery currently contains 25 different plugin events that are triggered at different points.
`onContentAfterDelete, onContentAfterSave, onContentPrepareData, onJoomAfterAddFavourite, onJoomAfterClearFavourites, onJoomAfterComment, onJoomAfterDisplayCatThumb, onJoomAfterDisplayDetailImage, onJoomAfterDisplayThumb, onJoomAfterRemoveFavourite, onJoomAfterReport, onJoomAfterSend2Friend, onJoomAfterTag, onJoomAfterUpload, onJoomAfterVote, onJoomBeforeComment, onJoomBeforeDownload, onJoomBeforeUpload, onJoomBeforeVote, onJoomBeforeZipDownload, onJoomCheckCaptcha, onJoomDisplayIcons, onJoomGetCaptcha, onJoomGetLastComments, onJoomSearch`

This PR currently adds another 10 events:
`onCategoryChangeState, onContentBeforeSave, onContentChangeState, onJoomAfterPrepareDisplayHTML, onJoomAfterRecreate, onJoomAfterRotate, onJoomAfterUnlockCat, onJoomBeforeSave, onJoomAfterCheckImage, onJoomAfterCheckCategory`

Maybe it would be helpful to write down all of them (as documentation for developers?) and specify when/for what purpose they are triggered and what can be done with them?

Elfangor93 commented:
> Maybe it would be helpful to write down all of them

Actually I am planning that in parallel to that PR. I am looking for a documentation on our website similar to [this one](https://documentation2051.rssing.com/chan-67472378/all_p1.html) but with all plugin events that are existent at this time and of course with that one added in this PR.

Actually you missed some:
`onContentPrepareForm, onJoomBeforeSendMessage, onJoomDisplayUser, onJoomGetSmileys, onJoomOpenImage, onJoomOpenImageGetName`

Elfangor93 commented:
> In my opinion, we should introduce new plugin events if it brings concrete benefits

Do I understand you correctly: You would introduce new plugin events only when they are used in a plugin. You dont want to introduce useful plugin events in advance?

MrMusic commented:
>Do I understand you correctly: You would introduce new plugin events only when they are used in a plugin. You dont want to introduce useful plugin events in advance?

No, useful plugin events can introduced certainly. 
The question is rather whether there is any 'idea' for what the plugin event could be used for.

Erftralle commented:
Maybe we can conclude a compromise.

We could focus on the new events **that we all think make sense**.
In my opinion, these are the ones that are necessary for the development of a Smart Search Plugin.

Elfangor93 commented:
I cleaned up this PR based on the compromiss stated in the joomgalleryfriends forum.
And I also updated the article to Plugin-Events on the joomgalleryfriends website such that the article corresponds to this PR.

Erftralle commented on administrator/components/com_joomgallery/models/categories.php:
```diff
@@ -721,6 +722,9 @@ public function publish($cid, $publish = 1, $task = 'publish')
     }
 
+    JPluginHelper::importPlugin('content');
+    $this->_mainframe->triggerEvent('onCategoryChangeState', array(_JOOM_OPTION.'.category', $cid, array('publish'=>$publish,'task'=>$task)));
```
According to the function signature, shouldn't just a `publish ` be passed instead of the associative array` array ('publish' => $ publish, 'task' => $ task)`?

Elfangor93 replied:
The associative array is for the plugin to differ from which task the event is coming from. If there would be just one task that calls the `publish` method in the `categories` model, it would make sense to pass just a `publish` to the plugin.

But there are actually two tasks calling the `publish` method in the `categories` model.
The `approve` method in the `categories` controller is calling the `publish` method too:
https://github.com/JoomGalleryfriends/JoomGallery/blob/3d2145f717a1234b1ce9d1b91375a0c724051825/administrator/components/com_joomgallery/controllers/categories.php#L143

Therefore there are two possibilities how the `onCategoryChangeState` event could be triggered. task could be `publish` or `approve`. Just passing a `publish` would not be enough information for the plugin. Or am I wrong?

Erftralle replied:
The trigger calls
```
	/**
	 * Smart Search change category state content method.
	 * Method is called when the state of the category to which the
	 * content item belongs is changed.
	 *
	 * @param   string   $extension  The extension whose category has been updated.
	 * @param   array    $pks        A list of primary key ids of the content that has changed state.
	 * @param   integer  $value      The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	public function onCategoryChangeState($extension, $pks, $value)
	{
		$dispatcher = JEventDispatcher::getInstance();
		JPluginHelper::importPlugin('finder');

		// Trigger the onFinderCategoryChangeState event.
		$dispatcher->trigger('onFinderCategoryChangeState', array($extension, $pks, $value));
	}
```
The signature says **_value_** should be an **_int_**!

Inside this function the trigger calls
```
	/**
	 * Method to update the item link information when the item category is
	 * changed. This is fired when the item category is published or unpublished
	 * from the list view.
	 * Event is triggered at the same time as the onCategoryChangeState event
	 *
	 * @param   string   $extension  The extension whose category has been updated.
	 * @param   array    $pks        A list of primary key ids of the content that has changed state.
	 * @param   integer  $value      The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	public function onFinderCategoryChangeState($extension, $pks, $value)
	{
		// Make sure we're handling joomgallery categories.
		if ($extension === 'com_joomgallery')
		{
			$this->categoryStateChange($pks, $value);
		}
	}
```
The signature says **_value_** should be an **_int_**! 
By the way the function `$this->categoryStateChange()` will  not be executed because you check for `if ($extension === 'com_joomgallery')` but you pass `'com_joomgallery.category'` when triggering. 

The function called inside is
```
	/**
	 * Method to update index data on category access level changes
	 *
	 * @param   array    $pks    A list of primary key ids of the content that has changed state.
	 * @param   integer  $value  The value of the state that the content has been changed to.
	 *
	 * @return  void
	 *
	 * @since   2.5
	 */
	protected function categoryStateChange($pks, $value)
	{
		/*
		 * The item's published state is tied to the category
		 * published state so we need to look up all published states
		 * before we change anything.
		 */
		foreach ($pks as $pk)
		{
			$query = clone $this->getStateQuery();
			$query->where('c.id = ' . (int) $pk);

			// Get the published states.
			$this->db->setQuery($query);
			$items = $this->db->loadObjectList();

			// Adjust the state for each item within the category.
			foreach ($items as $item)
			{
				// Translate the state.
				$temp = $this->translateState($item->state, $value);

				// Update the item.
				$this->change($item->id, 'state', $temp);

				// Reindex the item
				$this->reindex($item->id);
			}
		}
	}
```
The signature says **_value_** should be an **_int_**!

By the way, as far as I know approving a category can not be done anywhere in the JoomGallery GUI and if called directly via URL it would throw an error because there is no 'approved' field in the category database table. The code was probably prepared at some point in the past to approve categories, but it never got that far ;-).

Elfangor93 replied:
If we delete the "approving" in the code I am fine with just passing the integer state (0: unpublished, 1:published), to the event. I guess this will improve the compatibility of our code to the joomla core,

Erftralle replied:
I don't understand what the removal of the prepared but unused code has to do with the incorrect parameter transfer or with the  improvement of the compatibility of our code to the joomla core . If there was the function of approving categories, you couldn't just remove that code either.

You trigger a function of a Joomla plugin and according to the signature it expects an integer value when passing the parameters. This Joomla plugin triggers the `onFinderCategoryChangeState` function of all registered finder plugins, including those in the JoomFinderPlugin you provided. Your function also expects an integer value when passing this parameter. All other functions that transfer this parameter always expect an integer value (e.g. your Override of `translateState()` in the JoomFinderPlugin - you don't use this value in the function, which I don't understand either).

I know that PHP is not type safe and that you can pass any type when passing parameters, but this is bad programming style. In addition, the code will not work, because all functions interpret the "**_p_**" from "array ('publish' => $ publish, 'task' => $ task)" as integer value for the state and this value is always constant.

Elfangor93 reviewed (commented):
So what is your recommendation?

I see these different options:
1. Stick stricly to the Joomla core and pass only integers of 0 and 1. This will lead to serious problems in the JG plugins since JoomGallery knows more than two states (publish, approve, feature)
2. Add other integers for other states. Like 0&1 for publish, 2&3 for approve and 4&5 for feature. This could lead to problems for core Plugins since they only check the states 0 and 1.
3. Stay at the array approach which will solve the problem for JG plugins and will lead to problems in core plugins. And life with the uglyness of passing an array to a integer argument.

The first one is not an option to me as long as we have multiple states in categories and images.
The second one could be a compromiss and the third is still my prefered even having the uglyness of the array<>integer problem.

Erftralle commented on administrator/components/com_joomgallery/models/images.php:
```diff
@@ -591,6 +592,9 @@ public function publish($cid, $publish = 1, $task = 'publish')
     }
 
+    JPluginHelper::importPlugin('content');
+    $this->_mainframe->triggerEvent('onContentChangeState', array(_JOOM_OPTION.'.image', $cid, array('publish'=>$publish,'task'=>$task)));
```
> 1. Stick stricly to the Joomla core and pass only integers of 0 and 1. This will lead to serious problems in the JG plugins since JoomGallery knows more than two states (publish, approve, feature)

Afaik that's not true. The Joomla core knows also the state "Archived" for articles and it's value is "2". Which serious problems will occur and which JG plugin is using the event  "onCategoryChangeState" at the moment? The only one I know is your provided JoomFinderPlugin, which I' am testing together with this event. Unfortunately nothing happens with indexed images of a category if I publish or unpublish the catergory.

> 2. Add other integers for other states. Like 0&1 for publish, 2&3 for approve and 4&5 for feature. This could lead to problems for core Plugins since they only check the states 0 and 1.

Yes, that is an option. You could additionally use defined CONSTANTS to make the code more readable. Why should this lead to problems in the core plugins? Could you provide an example please. Every triggered function should and afaik does check with the parameter `extension` if it should do process the trigger.  And an **override** of `FinderIndexerAdapter::translateState()` should or could also be used.

> 3. Stay at the array approach which will solve the problem for JG plugins and will lead to problems in core plugins. And life with the uglyness of passing an array to a integer argument.

Most probably this statement has a typo. Do you meant "and will lead **not** to problems in core plugins." If yes than I have to ask you, which problems of which JG plugins do you mean and why should passing a wrong type **and** a wrong value to a Joomla core plugin lead to less problems than just passing a wrong integer value?

> Any ideas how we could solve this problem...

I don't see why this couldn't be solved with integers as well. Of course you have to adapt your code in JoomFinderPlugin accordingly.

Elfangor93 replied:
Oke, I changed it to option B.

The interpretation of the different integers are as the following: 
0: unpublished / 1: published / 2: archived / 3: not approved / 4: approved / 5: not featured / 6: featured

Elfangor93 commented:
Last commit was a change due to #90

Erftralle commented on administrator/components/com_joomgallery/models/category.php:
```diff
@@ -654,6 +679,15 @@ protected function batchAccess($value, $pks)
+        {
+          return false;
+        }
+
```
A question. Isn't it neccessary to additionally trigger 'onContentBeforeSave' for the JoomFinderPlugin here?

Erftralle commented on administrator/components/com_joomgallery/models/category.php:
```diff
@@ -948,6 +1010,14 @@ protected function batchMove($value, $pks)
+      {
+        return false;
+      }
+
```
Once again a question. Isn't it necessary to additionally trigger 'onContentBeforeSave' for the JoomFinderPlugin somewhere here in the function?

Erftralle commented on administrator/components/com_joomgallery/models/image.php:
```diff
@@ -363,6 +372,14 @@ public function store($data = null, $files = null, $params = null)
 
+      // Trigger Event onContentBeforeSave (Returnvalue: true or false)
+      JPluginHelper::importPlugin('content');
+		  $plugins = $this->_mainframe->triggerEvent('onContentBeforeSave', array(_JOOM_OPTION.'.image', &$row, true, $data));
```
Code style (use spaces instead of tabs)

Erftralle commented on administrator/components/com_joomgallery/models/images.php:
```diff
@@ -1006,6 +1049,10 @@ public function rotate()
 
+      // trigger Event "onJoomAfterRotate($files,$angle,$imgtypes)". Attached is an array with infos about the rotation
+      // files: path to the files; angle: the angle by which the images were rotated; imgtypes: 1(Thumbs and details), 2(All images)
+      $this->_mainframe->triggerEvent('onJoomAfterRotate', array(array('original'=>$orig,'detail'=>$img,'thumbnail'=>$thumb), $rotateImageAngle, $rotateimagetypes));
```
Code style (one space after the comma)

Erftralle commented on administrator/components/com_joomgallery/views/config/view.html.php:
```diff
@@ -214,6 +214,11 @@ public function display($tpl = null)
+    // Additional Tabs added by plugins
+    $this->event  = new stdClass();
+    $addTabs  = $this->_mainframe->triggerEvent('onJoomAfterDisplayTabs', array(_JOOM_OPTION.'.config',''));
+    $this->event->afterDisplayTabs = trim(implode('', $addTabs));
```
According to https://www.joomgalleryfriends.net/plugin-events.html#onJoomAfterDisplayTabs "$addTabs" has the type "string" but `implode('', $addTabs)` needs an array.

Erftralle commented on administrator/components/com_joomgallery/views/maintenance/view.html.php:
```diff
@@ -215,6 +215,11 @@ function display($tpl = null)
 
+    // Additional Tabs added by plugins
+    $this->event  = new stdClass();
+    $addTabs  = $this->_mainframe->triggerEvent('onJoomAfterDisplayTabs', array(_JOOM_OPTION.'.maintanance',$tab));
```
According to https://www.joomgalleryfriends.net/plugin-events.html#onJoomAfterDisplayTabs "$addTabs" has the type "string" but implode('', $addTabs) needs an array.

MrMusic commented:
I have tested this item ✅ successfully. Thank you very much.
